pkg/routes/middleware.go

```diff
@@ -25,19 +25,26 @@ import (
 	"github.com/zinclabs/zinc/pkg/auth"
 )
 
-func AuthMiddleware(c *gin.Context) {
-	// Get the Basic Authentication credentials
-	user, password, hasAuth := c.Request.BasicAuth()
-	if hasAuth {
-		if _, ok := auth.VerifyCredentials(user, password); ok {
-			c.Next()
+func AuthMiddleware(role string) func(c *gin.Context) {
```

Comment:
I'm not sure if it's a good idea to hardwire which role is allowed to access which resource like this. It's not very extensible this way. 

What if I do want normal users to create mapping for example ?

Maybe we could hide the role resource mapping behind some interface so later we can provide an implementation that uses a user defined mapping.

Reply:
Yes, we should rather provide the full rbac, but this is by far the easiest way

Reply:
But it will bite back when implementing a real rbac system.

Next to my remarks I also don't think it's correct to hardcode the role with the url. I would rather bind the url with a set of privileges and then again behind an interface make a simple mapping between role and privileges for now.

I'm not against hardcoding things as of now if it's extensible in the future.

Reply:
Handler sets a unique key that is checked against the role in `AuthMiddleware`.
like:
```
   r.GET("/api/index", AuthMiddleware("index.List"), index.List)
```

Reply:
Yes I like something like that much more, the `AuthMiddleware` receives a unique key or I would call it a privilege.

In `AuthMiddleware` I call upon an interface with one method `GetPrivilegesForRole(role string)`. Inside the implementation I would simply hold an in memory map that holds the mapping of privileges by roles. So in a later iteration this can be extended by a user defined mapping.

An alternative would be to make it part of the `User` where it will hold a set of `privileges`. Then you could add a method to the `User` `IsPrivileged(privilege string)` or `IsAuthorized(privilege string)`. Then you'll have to load the role privileges mapping when you load the user. This is in a way similar to how spring-security does it.

I must say I like the last option the best as you make access rights part of the `User` domain.

Reply:
I think this is a good start but hardcoding will indeed be problematic. I really like what jonas suggested:

> An alternative would be to make it part of the User where it will hold a set of privileges. Then you could add a method to the User IsPrivileged(privilege string) or IsAuthorized(privilege string). Then you'll have to load the role privileges mapping when you load the user. This is in a way similar to how spring-security does it